### ISSUE TITLE
How to detach the autocomplete from the `SearchBar`

### ISSUE THREAD

**Issue by hiagolcm** [vtex-apps/store-discussion#150]
**What are you trying to accomplish? Please describe.**  
Create a PR to detach autocomplete from `SearchBar` and keeps retro compatibility.

**What have you tried so far?**  
Currently, in store-components, it's not possible to create a custom autocomplete box while keeping the search-bar input.
If one wants to create a custom autocomplete, the whole `SearchBar` needs to be reimplemented.
To avoid this, I detached the autocomplete from the Search Bar. In this way, the `search-bar` interface becomes this:

```
// interfaces.json

"search-bar": {
    "component": "SearchBar",
    "required": ["results-list"]
}
"results-list": {
    "component": "ResultsList"
},
```

The component `ResultsList` is the default autocomplete used currently in the `SearchBar`.

Here are the diffs from my branch: [diffs](https://github.com/vtex-apps/store-components/compare/master...hiagolcm:feature/detach-autocomplete-from-search-input?expand=1)

Here is a workspace with the default autocomplete detached: [workspace](https://detachresultlist--biggy.myvtex.com/)

The problem with this approach is that as we add the `required` field to the `search-bar`, stores that have already declared a `search-bar` will break since they don't have the `blocks` field.
In other words, my solution doesn't have retro compatibility.

Is there a way to make these changes keeping the retro compatibility? If not, what is the best way? Fork the `SearchBar` and create a `SearchBarV2`?

**Comment by saviofelixmuniz** [comment]
Aren't we able to create an interface for a new search-bar instead of forking? The `search-bar` would remain: 
```
 "search-bar": {
    "component": "SearchBar"	  
  }
```
And there could be a `search-bar.v2` that would be like:

```
"search-bar.v2": {
    "component": "SearchBar"
    "required": ["results-list"]
  },
"results-list": {
   "component": "ResultsList"
}
```

The interface name could even be a bit better, something like: `autocomplete-search`

**Comment by victorhmp** [comment]
Hi @hiagolcm !
Just as @saviofelixmuniz said, the easiest way to do this is by simply extending the current `search-bar`, maybe creating `search-bar.v2`. We're doing this in a few other apps, such as `search-result` and, hopefully in the near future, `minicart`.
By doing this, you could even create a new `SearchBar` component and use it as the entry point for this new interface.
All of that would be a `minor` release of `vtex.store-components` and things should still work as expected :)

**Comment by victorhmp** [comment]
There's one more thing.
You guys could create a new app, `vtex.search-bar` and move the code from `store-components` to this new app. That would be pretty sweet since we're trying to move implementation away from `store-components`.
To make this work without breaking anything, you just need to list your app as a dependency at `store-components`, and keep the declaration for the interface `search-bar` where it is. 
This was done with the `image` component, you could use that as an example

**Comment by klzns** [comment]
Nice!

I would suggest using another name for `results-list`.

Once an app depends on another app, the interfaces of this app can be referenced using it name in `blocks.json`, for example, to use this interface one could type `results-list`. That is interfaces names become global in the context of this app.

Now imagine that we add another app dependency, it also defines an interface `results-list`. How to you reference it? The interface name can be disambiguated by specifying the app name `vtex.store-components@2.x:results-list` but we rarely do that and want to avoid.

The name `results-list` might be related to a search result page or something else.

I would go with a name more "namespaced" like "autocomplete-search-result-list".

**Comment by hiagolcm** [comment]
Guys, thanks for all replies!

I have chosen `results-list` as the name because the component responsible for the autocomplete in store-components is called `ResultsList`.  I don't like this name too and I supposed that this discussion would come up hahaha. @klzns   I think `autocomplete-search-result-list` looks nice!

I like the idea of extends the search-bar only in the interfaces as @saviofelixmuniz suggested, but how the `ExtensionPoint` will be treated in both cases (`search-bar` and `search-bar.v2`)?
I mean, I will add a `<ExtensionPoint id="results-list"`> where the autocomplete should appear, but `search-bar` doesn't know what `results-list` is.

**Comment by victorhmp** [comment]
What about `autocomplete-result-list`? I don't know if `search` is necessary (just a suggestion).
Oh! I see what you mean. When you declare a new interface `search-bar.v2`, you can create a new entry point for it, which means a new React component to be rendered by it. You could create a component named `SearchBarFlexible` (maybe a better name hahahah), for example, and define the new interface as:

```json
"search-bar.v2": {
    "component": "SearchBarFlexible"
    "required": ["results-list"]
  },
```

Then, this new SearchBarFlexible would be expecting `result-list` to be passed to it and rendered as an `<ExtensionPoint />` :)

**Comment by hiagolcm** [comment]
@victorhmp, @saviofelixmuniz has proposed to use the same entry point because we are trying to avoid to implement a new SearchBar with similar implementations of the old SearchBar.
Since there is no other way, I will try to keep `SearchBarFlexible`(or a better name) with as few duplicated code as possible.

About the name, I think both are nice. Let's discuss it in the PR I'm about to create. Maybe there is a better place for this kind of discussion.

If all suggestions have been proposed, feel free to close this issue.

**Comment by klzns** [comment]
Just for documentation purposes. 
The work is being done at:
- https://github.com/vtex-apps/store-components/pull/666
- https://github.com/vtex-apps/react-portal/pull/16
